### PR TITLE
PR: Fix several issues with the new LSP folding

### LINKED ISSUE #10777: KeyError when removing syntax error in editor
### What steps will reproduce the problem?

This error is generated when a syntax error is removed quickly in the editor. It seems caused by the new folding functionality.

### Traceback
```python-traceback
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/completion/plugin.py", line 120, in receive_response

Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/completion/plugin.py", line 120, in receive_response
    self._maybe_send_locked(req_id)
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/completion/plugin.py", line 161, in _maybe_send_locked
    send()
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/completion/plugin.py", line 143, in send
    self.gather_and_send(request_responses)
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/completion/plugin.py", line 212, in gather_and_send
    response_instance.handle_response(req_type, responses)
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/editor/widgets/codeeditor.py", line 928, in handle_response
    handler(params)
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/editor/widgets/codeeditor.py", line 1325, in handle_folding_range
    folding_panel.update_folding(ranges)
  File "/home/carlos/Projects/spyder/github-repo/spyder/plugins/editor/panels/codefolding.py", line 220, in update_folding
    folding_status[new_line] = self.folding_status[line]
KeyError: 297
```

## Versions

* Spyder version: 4.0.0.dev0 ddaf309bd
* Python version: 3.7.5
* Qt version: 5.9.6
* PyQt5 version: 5.9.2
* Operating System: Linux 4.13.16-041316-generic

### Dependencies

```
cloudpickle >=0.5.0          :  1.2.2 (OK)
pygments >=2.0               :  2.4.2 (OK)
qtconsole >=4.6.0            :  4.5.5 (NOK)
nbconvert >=4.0              :  5.6.1 (OK)
sphinx >=0.6.6               :  2.2.1 (OK)
pylint >=0.25                :  2.4.3 (OK)
psutil >=0.3                 :  5.6.3 (OK)
qtawesome >=0.5.7            :  0.6.0 (OK)
qtpy >=1.5.0                 :  1.9.0 (OK)
pickleshare >=0.4            :  0.7.5 (OK)
zmq >=17                     :  18.1.0 (OK)
chardet >=2.0.0              :  3.0.4 (OK)
numpydoc >=0.6.0             :  0.9.1 (OK)
spyder_kernels >=1.8.0;<2.0.0:  1.9.0.dev0 (OK)
qdarkstyle >=2.7             :  2.7 (OK)
atomicwrites >=1.2.0         :  1.3.0 (OK)
diff_match_patch >=20181111  :  20181111 (OK)
intervaltree                 :  None (OK)
watchdog                     :  None (OK)
keyring                      :  None (OK)
pexpect >=4.4.0              :  4.7.0 (OK)
pympler                      :  None (OK)
sympy >=0.7.3                :  1.4 (OK)
cython >=0.21                :  0.29.13 (OK)
IPython >=4.0                :  7.9.0 (OK)
matplotlib >=2.0.0           :  3.1.1 (OK)
pandas >=0.13.1              :  0.25.2 (OK)
numpy >=1.7                  :  1.17.3 (OK)
scipy >=0.17.0               :  1.3.1 (OK)
pyls >=0.31.0;<0.32.0        :  0.31.1 (OK)
xdg >=0.26                   :  0.26 (OK)
rtree >=0.8.3                :  0.8.3 (OK)
```

### PR DESCRIPTION
**Pull request by ccordoba12**
- This fixes several KeyError's that are very frequent with our new folding implementation and leads to nasty paint event errors too.
- We also request folding now after linting results arrive from the server, so it's updated for all edition cases where linting is requested.
- Fnally, indent guides are computed only when users enable them.

Fixes #10777.

### CODE REVIEW

File: spyder/plugins/editor/panels/codefolding.py [hunk 1]
```diff
@@ -213,11 +213,14 @@ def update_folding(self, ranges):
...
-                offset = prev_offsets[line]
-                new_offset = differ.diff_xIndex(diff, offset)
-                new_line = current_lines[new_offset]
-                if new_line in self.folding_regions:
-                    folding_status[new_line] = self.folding_status[line]
+                try:
+                    offset = prev_offsets[line]
+                    new_offset = differ.diff_xIndex(diff, offset)
+                    new_line = current_lines[new_offset]
+                    if new_line in self.folding_regions:
```
Comment (goanpeca) [inline spyder/plugins/editor/panels/codefolding.py]: Why not something like:
```python
                    offset = prev_offsets[line]
                    new_offset = differ.diff_xIndex(diff, offset)
                    new_line = current_lines[new_offset]
                    if new_line in self.folding_regions and line in self.folding_status[line]:
                        folding_status[new_line] = self.folding_status[line]
```

or

```python
                try:
                    if new_line in self.folding_regions: 
...
```

The problem line is only `folding_status[new_line] = self.folding_status[line]`

**Comment by ccordoba12** [comment]
@andfoy is going to take a look at this now.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Let me know if/when you want me to test it. Thanks!

**Comment by ccordoba12** [comment]
@andfoy, I'm going to merge this one as is. Later we can see how to fix the problems with folding not being updated under certain circumstances.

**Comment by ccordoba12** [comment]
@CAM-Gerlach, this is ready for you to test it.

**Comment by ccordoba12** [comment]
@CAM-Gerlach, I'm going to merge this because it fixes several ugly errors in master. Please let us know if it helps with the lagging issues introduced by indent guides on issue #8864.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 I initially thought this functionality was still broken, but it turns out that this only enables indent guides in the top-left most Editor pane, and indent guides don't work in any other pane.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Should I open a new issue, since indent guides are still broken in any other pane but the first?

**Comment by ccordoba12** [comment]
Sure, please do.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Done as #10900 . Its a breaking, very late regression (rc3) in a major new feature introduced all the way back in beta 1, so it should probably be a blocker on the final release IMO.

**Pull request merged**